### PR TITLE
Bugfix for Nesterov Momentum Optimier

### PR DESCRIPTION
**Pull request by albi3ro**
This bug was discovered through [the "Variational Classifier" demo](https://github.com/PennyLaneAI/qml/runs/6085549063?check_suite_focus=true#step:8:2432).

In the Nesterov Momentum Optimizer, there was confusion between the index of an argument in the total arguments list and the index in the *trainable-only* arguments list.  Non-trainable arguments were shifted using the gradient of a trainable argument after them in the list. Since `BasisState` has to have zero and one only, this shifting caused execution to fail.

### CODE REVIEW

File: tests/optimize/test_optimize.py [hunk 1]
```diff
@@ -115,17 +115,21 @@ def test_nontrainable_data(self, opt):
         """Check non-trainable argument does not get updated"""
 
         def func(a, b, c, d):
+            if not qml.math.allclose(b, 1.0):
+                # Tests Nesterov momentum optimizer only shifts arguments of the trainable parameters
+                raise ValueError("Only accepts a b of 1.0.")
             return a * b * c * d
 
-        a = np.array(1.0, requires_grad=True)
+        a = np.array(0.1, requires_grad=True)
```
Comment (antalszava) [inline tests/optimize/test_optimize.py:123]: @albi3ro it seems that the change in this test doesn't raise an error similar to the one in the demo. Is that correct? What could be a good test case to catch the edge case from the demo?

Comment (albi3ro) [reply tests/optimize/test_optimize.py:123]: For me, this test failed on the release candidate branch.

Comment (antalszava) [reply tests/optimize/test_optimize.py:123]: That's the trick, thanks for pointing it out!

Any reasons for going with an error instead of an `assert`?

File: tests/optimize/test_optimize.py [hunk 2]
```diff
@@ -115,17 +115,21 @@ def test_nontrainable_data(self, opt):
         """Check non-trainable argument does not get updated"""
 
         def func(a, b, c, d):
+            if not qml.math.allclose(b, 1.0):
+                # Tests Nesterov momentum optimizer only shifts arguments of the trainable parameters
+                raise ValueError("Only accepts a b of 1.0.")
```
Comment (albi3ro) [inline tests/optimize/test_optimize.py:120]:
```suggestion
            assert qml.math.allclose(b, 1.0)
```

**Review by antalszava — APPROVED** [review]
Looks good, thank you for getting this through so quick! :rocket: 

The demo executes well locally. :partying_face:

**Comment by antalszava** [comment]
Overriding the Codefactor check: after checking locally, no changes in the PR have affected linting in an unwanted way.
